--- shadow-4.2.1/libmisc/loginprompt.c
+++ shadow-4.2.1/libmisc/loginprompt.c
@@ -98,8 +98,7 @@ void login_prompt (const char *prompt, char *name, int namesize)
 				(void) fclose (fp);
 			}
 		}
-		(void) gethostname (buf, sizeof buf);
-		printf (prompt, buf);
+		printf (prompt);
 		(void) fflush (stdout);
 	}
 
--- shadow-4.2.1/src/login.c
+++ shadow-4.2.1/src/login.c
@@ -926,7 +926,7 @@ int main (int argc, char **argv)
 			preauth_flag = false;
 			username = xmalloc (USER_NAME_MAX_LENGTH + 1);
 			username[USER_NAME_MAX_LENGTH] = '\0';
-			login_prompt (_("\n%s login: "), username, USER_NAME_MAX_LENGTH);
+			login_prompt (_("\nlogin: "), username, USER_NAME_MAX_LENGTH);
 
 			if ('\0' == username[0]) {
 				/* Prompt for a new login */
