--- busybox-1.23.1/libbb/login.c
+++ busybox-1.23.1/libbb/login.c
@@ -13,7 +13,7 @@
 /* After libbb.h, since it needs sys/types.h on some systems */
 #include <sys/utsname.h>
 
-#define LOGIN " login: "
+#define LOGIN "login: "
 
 static const char fmtstr_d[] ALIGN1 = "%A, %d %B %Y";
 
@@ -89,12 +89,8 @@ void FAST_FUNC print_login_issue(const char *issue_file, const char *tty)
 
 void FAST_FUNC print_login_prompt(void)
 {
-	char *hostname = safe_gethostname();
-
-	fputs(hostname, stdout);
 	fputs(LOGIN, stdout);
 	fflush_all();
-	free(hostname);
 }
 
 /* Clear dangerous stuff, set PATH */